### DISCUSSION TITLE
Feedback and questions about incremental build

### DISCUSSION THREAD

**Discussion by belaczek** [nrwl/nx#12600] [General]
Following is a write-up of my struggles adopting incremental build in our monorepo. There are a couple of questions for which I wasn't able to find an answer.

**Our setup:** NX monorepo with 2 react apps (powered by `@nrwl/webpack` executors) and 5 shared libraries (buildable using `@nrwl/rollup`, some non-buildable).

Now to issues I have experienced so far:

1. There's no clear statement about what exactly has to be done to enable incremental build. I understand the build task needs to have a dependency on the build of its dependencies. But that alone won't ensure the built assets of libraries will be used instead of the library source code. Eventually, I found in this tutorial on [how to make the build work in Angular](https://nx.dev/recipes/other/setup-incremental-builds-angular#adjust-the-application-executor) that  there's this `buildLibsFromSource: true` parameter, which does the trick. 
Is this all that has to be done to make the incremental build work?
2. I noticed the `buildLibsFromSource` param only exists on `@nrwl/webpack:webpack` executor. How can I control this elsewhere, for example in `@nrwl/rollup` or `@nrwl/esbuild`?
3. For a long time I struggled to make the project compile, only to find out [here](https://github.com/nrwl/nx/issues/11374), that it is caused by having custom paths override defined in a project level `tsconfig`. This would be good to have mentioned in the documentation.
4. When I got the incremental build running (or at least I hope), it turned out that there was hardly any benefit in terms of compile speed performance. The app alone has about the same compile time, no matter if libraries are prebuilt or not.
I understand that the incremental build is achieved by swapping tsconfig paths to point to the built artifacts in the dist folder. However, it seems like these assets are still being processed again by `babel-loader` anyway. I'd rather expect them to be handled the same way `node_modules` are: bundled but not transpilled. Is that the wrong expectation? Or is there something else happening in the background?

I hope it doesn't sound too negative. The goal of this post isn't to complain, but rather just to provide notes from real-life experience and hopefully get some answers.

**Comment by lukascivil** [comment]
I'm having the same problem. I'm still waiting for the response from the maintainers. Anyway, here's a possible way, which is different from NX case, but can be the light for incremental builds with the help of an orchestrator, Turbopack/repo.

**Comment by viniciusbsneto** [comment]
Hello, @belaczek! I'm having a hard time at trying to understand this incremental build feature too. That Angular + webpack-browser executor example got me really confused. Especially for the `buildLibsFromSource` boolean. It's not clear if this is some additional configuration for Angular projects or an outdated documentation. It's not clear if there's something else we should be doing whether we're working with another bundler. Anyways I've managed to get some half-answers on this [thread](https://nrwlcommunity.slack.com/archives/CMFKWPU6Q/p1685375664524579) from Nx official Slack channel. I'm still really confused but I hope it helps you somehow.

**Comment by Chanki-Min** [comment]
I have same question with Next.js projects.

By looking with-nx code (which wraps next.config.js ), buildLisFromsSource flag is used to get buildable lib's locaiton and saves to block-scoped variable ([code](https://github.com/nrwl/nx/blob/57247e0b81e032fcdfac6e53e02ccd0bc29e9140/packages/next/plugins/with-nx.ts#L187)). But this variable (dependencies) is never used after allocation.
In contrast, in angular build code (which introduced @ nx's documentation) uses buildLisFromsSource flag in build step to make temporary ts config ([code](https://github.com/search?q=repo%3Anrwl%2Fnx+buildLibsFromSource+path%3Apackages%2Fangular&type=code))

I don't understand much of NX's incremental build, but it seems like NX's nextjs build is not doing any meaningful action to support incremental build. Am i wrong?  Does Nx supports incremental build in Next.js?

---
Moreover, If Nx supports somewhat incremental build with Next.js, does it gives us faster build time?

in with-nx , Nx forces to transpile buildable libs with Next.js's bundler (whitch is babel or swc based on user configuration to support pure-esm buildable libraris maybe? ([code](https://github.com/nrwl/nx/blob/57247e0b81e032fcdfac6e53e02ccd0bc29e9140/packages/next/plugins/with-nx.ts#L205))

By then, if Nx supports incremental build with Next.js, build speed is not going to be faster than normal build?

**Comment by danbeo95** [comment]
Anyone apply this for production

**Comment by JoelBonetR** [comment]
Couldn't find a way to get incremental builds in place either. Using Next.js and a bunch of TS and React libs.

**Comment by RoeeIfergan** [comment]
Has anyone found a solution for TS packages? using NX 21.6.4

My NX Monorepo has a client (react 18), server (NestJS) & multiple packages.
Some packages are UI librarys (EJS only), some are server only (CJS - NestJS requires CJS) and the reset are needed by both the client and server (need to be compiled to both EJS and CJS).

My current configuration (dev mode):
client: compiled and served by vite dev server)
server: compiled with nx:run-commands (webpack-cli build) and run with the @nx/js:node executor, the default executors for a NestJS app in NX
Client packages (EJS only) - compiled with SWC using @nx/js:lib's config
All other packages are compiled with vite library mode using @nx/js:lib's config (edited to build to both EJS & CJS)


I've gotten this to work in "laboratory conditions" but when every a dev experiences slowness on builds, it seems like were running into a race condition that crashes one of the builds (usually the server build) and we can't figure out why. 

Would love for some help debugging the issue.
If anyone can refrence a working incremnetal build monorepo example (For ts packages) that would be great!
Recommendations/Tips for incremental builds
